src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/RuntimeAsyncRewriter.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+internal sealed class RuntimeAsyncRewriter : BoundTreeRewriterWithStackGuard
+{
+    public static BoundStatement Rewrite(
+        BoundStatement node,
+        MethodSymbol method,
+        TypeCompilationState compilationState,
+        BindingDiagnosticBag diagnostics)
+    {
+        if (!method.IsAsync)
+        {
+            return node;
+        }
+
+        var rewriter = new RuntimeAsyncRewriter(compilationState.Compilation, new SyntheticBoundNodeFactory(method, node.Syntax, compilationState, diagnostics));
+        return (BoundStatement)rewriter.Visit(node);
+    }
+
+    private readonly CSharpCompilation _compilation;
+    private readonly SyntheticBoundNodeFactory _factory;
+
+    private RuntimeAsyncRewriter(CSharpCompilation compilation, SyntheticBoundNodeFactory factory)
+    {
+        _compilation = compilation;
+        _factory = factory;
+    }
+
+    private NamedTypeSymbol Task
+    {
+        get => field ??= _compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task);
+    } = null!;
+
+    private NamedTypeSymbol TaskT
+    {
+        get => field ??= _compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task_T);
+    } = null!;
+
+    private NamedTypeSymbol ValueTask
+    {
+        get => field ??= _compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_ValueTask);
+    } = null!;
+
+    private NamedTypeSymbol ValueTaskT
+    {
+        get => field ??= _compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_ValueTask_T);
+    } = null!;
+
+    public BoundExpression VisitExpression(BoundExpression node)
+    {
+        var result = Visit(node);
+        Debug.Assert(result is BoundExpression);
```

Comment:
nit: The assertion doesn't seem useful since we have a cast below (will throw anyways)